src/sql/postgres.zig

```diff
@@ -0,0 +1,3960 @@
+            globalObject: *JSC.JSGlobalObject,
+            value: *Data,
+        ) anyerror!JSC.JSValue {
+            defer value.deinit();
+            return JSC.JSValue.fromDateString(globalObject, value.sliceZ().ptr);
+        }
+    };
+
+    pub const bytea = struct {
+        pub const to = 17;
+        pub const from = [_]short{17};
+
+        pub fn toJS(
+            globalObject: *JSC.JSGlobalObject,
+            value: *Data,
+        ) anyerror!JSC.JSValue {
+            defer value.deinit();
+
+            // var slice = value.slice()[@min(1, value.len)..];
+            // _ = slice;
+            return JSC.JSValue.createBuffer(globalObject, value.slice(), null);
+        }
+    };
+};
+
+const Socket = uws.AnySocket;
+const PreparedStatementsMap = std.HashMapUnmanaged(u64, *PostgresSQLStatement, bun.IdentityContext(u64), 80);
+
+pub const PostgresSQLContext = struct {
+    tcp: ?*uws.SocketContext = null,
+
+    onQueryResolveFn: JSC.Strong = .{},
+    onQueryRejectFn: JSC.Strong = .{},
+
+    pub fn init(globalObject: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) callconv(.C) JSC.JSValue {
+        var ctx = &globalObject.bunVM().rareData().postgresql_context;
+        ctx.onQueryResolveFn.set(globalObject, callframe.argument(0));
+        ctx.onQueryRejectFn.set(globalObject, callframe.argument(1));
+
+        return .undefined;
+    }
+
+    comptime {
+        if (!JSC.is_bindgen) {
+            @export(init, .{
+                .name = "PostgresSQLContext__init",
+            });
+        }
+    }
+};
+
+pub const PostgresSQLQuery = struct {
+    statement: ?*PostgresSQLStatement = null,
+    query: bun.String = bun.String.empty,
+    cursor_name: bun.String = bun.String.empty,
+    thisValue: JSC.JSValue = .undefined,
+    target: JSC.Strong = JSC.Strong.init(),
+    status: Status = Status.pending,
+    is_done: bool = false,
+    ref_count: std.atomic.Value(u32) = std.atomic.Value(u32).init(1),
+    binary: bool = false,
+
+    pub usingnamespace JSC.Codegen.JSPostgresSQLQuery;
+
+    pub const Status = enum(u8) {
+        pending,
+        written,
+        running,
+        binding,
+        success,
+        fail,
+
+        pub fn isRunning(this: Status) bool {
+            return this == .running or this == .binding;
+        }
+    };
+
+    pub fn hasPendingActivity(this: *@This()) callconv(.C) bool {
+        return this.ref_count.load(.Monotonic) > 1;
+    }
+
+    pub fn deinit(this: *@This()) void {
+        if (this.statement) |statement| {
+            statement.deref();
+        }
+        this.query.deref();
+        this.cursor_name.deref();
+        this.target.deinit();
+
+        bun.default_allocator.destroy(this);
+    }
+
+    pub fn finalize(this: *@This()) callconv(.C) void {
+        debug("PostgresSQLQuery finalize", .{});
+        this.thisValue = .zero;
+        this.deref();
+    }
+
+    pub fn deref(this: *@This()) void {
+        const ref_count = this.ref_count.fetchSub(1, .Monotonic);
```

Comment:
```suggestion
        const ref_count = this.ref_count.fetchSub(1, .Monotonic);
        bun.assert(ref_count > 0);
```

Reply:
this will compile away to nothing because ref_count is an unsigned integer